### ISSUE TITLE
Directory-Zip-Entries missing in generated JAR

### ISSUE THREAD

**Issue by tristanlins** [webjars/webjars#1863]
The newly generated webjar for [@saeris/vue-spinners](https://search.maven.org/artifact/org.webjars.npm/saeris__vue-spinners/1.0.8/jar) misses the Directory-Zip-Entries.

When listing the entries with unzip, they are missing:

```
$ unzip -l saeris__vue-spinners-1.0.8.jar
Archive:  saeris__vue-spinners-1.0.8.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
     1518  2019-08-05 10:10   META-INF/maven/org.webjars.npm/saeris__vue-spinners/pom.xml
      104  2019-08-05 10:10   META-INF/maven/org.webjars.npm/saeris__vue-spinners/pom.properties
        0  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8//
    17437  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8/.eslintrc.js
       53  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8/babel.config.js
        0  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8/coverage/
        0  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8/example/
      531  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8/jest.config.js
        0  2019-08-05 10:10   META-INF/resources/webjars/saeris__vue-spinners/1.0.8/lib/
...
```

Compared to another webjars, it should look like this:

```
$ unzip -l vue-2.6.8.jar
Archive:  vue-2.6.8.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2019-03-01 21:28   META-INF/
        0  2019-03-01 21:28   META-INF/maven/
        0  2019-03-01 21:28   META-INF/maven/org.webjars.npm/
        0  2019-03-01 21:28   META-INF/maven/org.webjars.npm/vue/
     1427  2019-03-01 21:28   META-INF/maven/org.webjars.npm/vue/pom.xml
       87  2019-03-01 21:28   META-INF/maven/org.webjars.npm/vue/pom.properties
        0  2019-03-01 21:28   META-INF/
        0  2019-03-01 21:28   META-INF/resources/
        0  2019-03-01 21:28   META-INF/resources/webjars/
        0  2019-03-01 21:28   META-INF/resources/webjars/vue/
        0  2019-03-01 21:28   META-INF/resources/webjars/vue/2.6.8/
     5547  2019-03-01 21:28   META-INF/resources/webjars/vue/2.6.8/package.json
     1091  2019-03-01 21:28   META-INF/resources/webjars/vue/2.6.8/LICENSE
    17643  2019-03-01 21:28   META-INF/resources/webjars/vue/2.6.8/README.md
...
```

Because the entries are missing, the URLClassLoader / Tomcat WebappClassLoader fails to find the jar when query for resources by directory path (thats what the webjar locator do):

```java
final URLClassLoader loader = new URLClassLoader(
    new URL[]{
        new File("./vue-2.6.8.jar").toURL(),
        new File("./saeris__vue-spinners-1.0.8.jar").toURL()
    }
);
final ArrayList<URL> list = Collections.list(loader.getResources("META-INF"));
list.stream()
    .map(URL::toString)
    .filter(s -> s.contains("vue"))
    .sorted()
    .forEach(System.out::println);
```

Only the `vue-2.6.8.jar` is found. The the `saeris__vue-spinners-1.0.8.jar` is missing.

It is easy to patch the jar locally:

```
$ mkdir /tmp/jar
$ cd /tmp/jar
$ unzip -x ../saeris__vue-spinners-1.0.8.jar
Archive:  saeris__vue-spinners-1.0.8.jar
  inflating: META-INF/maven/org.webjars.npm/saeris__vue-spinners/pom.xml  
  inflating: META-INF/maven/org.webjars.npm/saeris__vue-spinners/pom.properties  
   creating: META-INF/resources/webjars/saeris__vue-spinners/1.0.8/
...
$ zip -u -r ../saeris__vue-spinners-1.0.8.jar .
  adding: META-INF/ (stored 0%)
  adding: META-INF/resources/ (stored 0%)
  adding: META-INF/resources/webjars/ (stored 0%)
  adding: META-INF/resources/webjars/saeris__vue-spinners/ (stored 0%)
  adding: META-INF/resources/webjars/saeris__vue-spinners/1.0.8/ (stored 0%)
  adding: META-INF/maven/ (stored 0%)
  adding: META-INF/maven/org.webjars.npm/ (stored 0%)
  adding: META-INF/maven/org.webjars.npm/saeris__vue-spinners/ (stored 0%)
```

As you can see, the zip added only the missing directory entries.

----

Environment Info

```
$ java -version                                                                                                                                             
openjdk version "12.0.1" 2019-04-16
OpenJDK Runtime Environment AdoptOpenJDK (build 12.0.1+12)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 12.0.1+12, mixed mode, sharing)

$ uname -a
Linux thinkpad 5.0.0-23-generic #24~18.04.1-Ubuntu SMP Mon Jul 29 16:12:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ lsb_release -a
No LSB modules are available.
Distributor ID: neon
Description:    KDE neon User Edition 5.16
Release:        18.04
Codename:       bionic
```

**Comment by jamesward** [comment]
I recently changed this because I'd read that the directory entries didn't matter.  But I guess if something is looking for an entry of `META-INF` then it will fail with JARs created after this change.

But I think you are right that webjars-locator filters out possible jars to look for WebJars in by doing a `loader.getResources("META-INF/resources/webjars")` query.  And now that won't pass.

Option 1) Add the dirs back in leaving some JARs created in the past few days broken
Option 2) Fix webjars-locator to not depend on that path

I'm kinda leaning towards option 2 if there isn't too much runtime overhead.  WDYT?

**Comment by tristanlins** [comment]
I'm not sure if option 2 is technically possible.
How will you find the existing webjars without any "pointer" to identify them?
runtime overhead can always be mitigated by caching ;-)

**Comment by tristanlins** [comment]
I thought again about Option 2 and I have a clear preference for Option 1.
From a ZIP point of view, a directory entry may not be really relevant because all (un)zip programs can handle it.
But from the point of view of the Java ClassLoader, a ZIP or JAR is regarded as a kind of file system. To refrain from the entries makes browsing the Webjars using ClassLoader impossible.

Here you should not consider this from the point of view of a simple ZIP, but from the point of view of the Java ClassLoader. And this obviously expects the entries to exist.

**Comment by manolo** [comment]
Directory listing in `.jar` is useful when for instance extracting all resources matching a pattern in your classpath `.jar` files when you need them for web packing, etc. 

It's a common practice to get all `META-INF/my-app`, then filter the `.jar` files which contains that folder, then extract the content.

In addition it is a common practice in `.jar`s to contain directory listing, since the standard `jar -cf` and `zip -r` cli tools and the jar-maven plugin do it in this way by default, so breaking this imo is not a good idea. I don't think there is any improvement in size, or whatever by not having the folder entries.

**Comment by phucnh** [comment]
> Option 1) Add the dirs back in leaving some JARs created in the past few days broken
> Option 2) Fix webjars-locator to not depend on that path

The missing directory entries jar files are increasing. So, I think the `Option 1)` should be done first.

About `Option 2)`
After some investigation, seems that it is impossible to get the resources from jar file without directory entries by `java.lang.ClassLoader`.

The alternative way is unzip the jar file by `java.util.jar.JarEntry`, then do filtering manually. 
(With some performance considerations as @tristanlins said)

The example code is as follows:

```scala
package com.example

import java.io.File
import java.net.{URL, URLClassLoader}
import java.util.jar.{JarEntry, JarFile}
import scala.collection.JavaConverters._

object JarEntryTest extends App {
  val url = new URL("file:/Users/a13942/.ivy2/cache/org.webjars.npm/monaco-editor/jars/monaco-editor-0.17.1.jar").toURI()
  val file = new File(url)
  val jarFile = new JarFile(file)

  val enum = jarFile.entries()

  println("The URL:")
  enum.asScala.to[List].foreach(println)
}
```

Output:
```console
The URL:
META-INF/maven/org.webjars.npm/monaco-editor/pom.xml
META-INF/maven/org.webjars.npm/monaco-editor/pom.properties
META-INF/resources/webjars/monaco-editor/0.17.1/CHANGELOG.md
META-INF/resources/webjars/monaco-editor/0.17.1/monaco.d.ts
META-INF/resources/webjars/monaco-editor/0.17.1/dev/bundleInfo.json
META-INF/resources/webjars/monaco-editor/0.17.1/dev/nls.metadata.json
META-INF/resources/webjars/monaco-editor/0.17.1/dev/vs/base/worker/workerMain.js
...
```

WDYT?

**Comment by jamesward** [comment]
Yeah, it seems like adding in directory entries for:
```
META-INF/
META-INF/resources/
META-INF/resources/webjars/
META-INF/resources/webjars/[WEBJAR_ID]/
META-INF/resources/webjars/[WEBJAR_ID]/[VERSION]/
```
Seems like a good way to provide optimizations for ClassLoader filtering.

Related, I've been switching out `webjars-locator-core`'s location underpinnings to use https://github.com/classgraph/classgraph/ which doesn't depend on these dir entries.

**Comment by jamesward** [comment]
The WebJar creator is now adding dir entries as described above:
```
$ jar -tvf saeris__vue-spinners-1.0.8-2.jar 
     0 Wed Aug 14 16:07:22 MDT 2019 META-INF/
     0 Wed Aug 14 16:07:22 MDT 2019 META-INF/resources/
     0 Wed Aug 14 16:07:22 MDT 2019 META-INF/resources/webjars/
     0 Wed Aug 14 16:07:22 MDT 2019 META-INF/resources/webjars/saeris__vue-spinners/
     0 Wed Aug 14 16:07:22 MDT 2019 META-INF/resources/webjars/saeris__vue-spinners/1.0.8-2/
```

From: https://repo1.maven.org/maven2/org/webjars/npm/saeris__vue-spinners/1.0.8-2/saeris__vue-spinners-1.0.8-2.jar

**Comment by tristanlins** [comment]
Thanks alot 🤗

**Comment by phucnh** [comment]
Thank you so much

**Comment by spencerhakim** [comment]
I'm not sure if only SBT has issues with the jars produced over the last 2 weeks or if other build tools are affected as well, but is there any chance that the affected artifacts can be deleted or the jars modified to include the directory entries?

**Comment by jamesward** [comment]
Unfortunately the releases are immutable.  They should work still, but might require a newer webjars-locator or another way of discovering the WebJars.  If you have problems finding a workaround, let me know.

**Comment by spencerhakim** [comment]
Thanks for pointing me in that direction . I tried adding the following to my `project/plugins.sbt` but it didn't fix the issue.

```scala
dependencyOverrides ++= Set(
    "org.webjars" % "webjars-locator" % "0.36",
    "org.webjars" % "webjars-locator-core" % "0.38"
)
```

At the moment I'm pinning affected dependencies to older versions manually, but that's obviously not an ideal solution. Do you happen to know if `webjars-locator` is responsible for extracting the jars into `target/web/node-modules` and `target/web/web-modules`, or is that `sbt-web`'s job?

**Comment by jamesward** [comment]
I'm still testing the new release so it's not quite ready.  And yes, `sbt-web` does use `webjars-locator-core` so I'm testing it too.

**Comment by jamesward** [comment]
I've released `webjars-locator` `0.37` but I need to get a change into `sbt-web` for it to work there.

**Comment by mkurz** [comment]
FYI: Upcoming sbt-web 1.5.0 depends on latest webjars-locator. I already released 1.5.0-M1 for testing purposes.
